plugin/entryBase.go

```diff
@@ -39,18 +39,17 @@ type EntryBase struct {
 	attr               EntryAttributes
 	slashReplacerCh    rune
 	// washID represents the entry's wash ID. It is set in CachedList.
-	washID string
-	ttl    [3]time.Duration
+	washID             string
+	ttl                [3]time.Duration
+	_shortType         string
+	singleton          bool
 }
 
 // NewEntry creates a new entry
-func NewEntry(name string) EntryBase {
-	if name == "" {
-		panic("plugin.NewEntry: received an empty name")
-	}
-
+//
+// TODO: Update docs with the changes to NewEntry
```

Comment:
TODOs like this make it really hard to review a work by individual commits. If you're going to resolve the TODO in a single body of work, I'd rather it amend the commit where the TODO was introduced.

Also, there's no explanation about why you're fundamentally changing the interaction with plugin.EntryBase. Name feels like a fundamental part of an entry, so why is it no longer required?

Reply:
I haven't reviewed all the changes to specific plugins yet because this question feels fairly fundamental to the PR.

Reply:
Main reason for the change is because Go's metaprogramming sucks (see my comment in https://github.com/puppetlabs/wash/pull/321#discussion_r292224305). TLDR is that the schema stuff is equivalent to class-level annotations in other languages, but Go has no way of doing that easily and in a way that isn't error-prone/fragile. So, the somewhat "hacky" way for Wash to figure out what an Entry looks like is to receive a "stub"/"template" object containing all the class-level configuration. That "stub/template" object is also the basis for creating specific instances of those objects (e.g. like EC2 instances, containers, S3 buckets, etc.) -- that's why the plugin-specific changes have things like `containerTemplate()`, `ec2InstanceTemplate()` (recommend reviewing the Docker changes).

`NewEntry` established an initial EntryBase template that you can then extend with your plugin-specific config like caching. So, it should really be `NewEntryTemplate()`. I can change its name to reflect that.

Biggest fundamental change for core plugin design is you have to specify entry schemas, hence why I made `ChildSchemas` part of the Parent interface. I also wanted a way to make it clear, at the type-level, that `ChildSchemas` returned schemas, which is why I didn't name it `ChildTemplates` (that's why `plugin.ChildSchemas` is there so you can at least do everything in one-two lines of code while getting that type-level safety). So the pattern for core plugin authors now is (1) Set-up a template object for your entry that you will then pass-into the parent's ChildSchema method; (2) Extend that template-object with instance-specific info like the name, attributes, etc.). Hence why there's `containerTemplate()`, then `newContainer()` in e.g. the Docker plugin.

Reply:
Could we have separate `NewEntry` and `NewEntryTemplate` functions, where `NewEntry` still takes the name as an argument? That would help distinguish between different types of use.

Reply:
Hmm, maybe. It'd be weird explaining why `newContainer()` uses a different EntryBase object than `containerTemplate()` though (when it shouldn't, `containerTemplate`'s EntryBase is the right EntryBase object).

What's wrong with `SetName`?

Reply:
I do think having a `NewEntryTemplate`/`EntryTemplate` method makes sense, just not sure that we should have `NewEntry` as well.

Reply:
It feels like conflating two concepts: the template, vs a named entry. And I don't care for the extra method call for something that's going to be used almost every time. An Entry has a name, while you're re-using it for something that's entry-like (a template of an entry).

Reply:
In my mind, `NewEntryTemplate` creates an a skeleton of the entry type. When you name it, it becomes an instance, and I feel like `NewEntry` represents that concept of a named entry.